matchers/php/todomatcher.go

```diff
@@ -2,37 +2,83 @@ package php
 
 import (
 	"regexp"
+	"sync"
 
 	"github.com/preslavmihaylov/todocheck/matchers/errors"
 )
 
-var singleLineTodoPattern = regexp.MustCompile("^\\s*//.*TODO")
-var singleLineValidTodoPattern = regexp.MustCompile("^\\s*// TODO (#?[a-zA-Z0-9\\-]+):.*")
+var lock sync.Mutex
 
-var singleLineScriptTodoPattern = regexp.MustCompile("^\\s*#.*TODO")
-var singleLineScriptValidTodoPattern = regexp.MustCompile("^\\s*# TODO (#?[a-zA-Z0-9\\-]+):.*")
+var singleLineTodoPattern *regexp.Regexp
+var singleLineValidTodoPattern *regexp.Regexp
 
-var multiLineTodoPattern = regexp.MustCompile("(?s)^\\s*/\\*.*TODO")
-var multiLineValidTodoPattern = regexp.MustCompile("(?s)^\\s*/\\*.*TODO (#?[a-zA-Z0-9\\-]+):.*")
+var singleLineScriptTodoPattern *regexp.Regexp
+var singleLineScriptValidTodoPattern *regexp.Regexp
+
+var multiLineTodoPattern *regexp.Regexp
+var multiLineValidTodoPattern *regexp.Regexp
```

Comment:
I guess these global variables are used so that we don't initialise all these regexps every time a new matcher for a file is created, which happens for every file.

Instead of making this optimization here, you can do it in the matcher factories. There, instead of initializing a matcher every time, initialise them as fields of the `matcherFactory` and reuse the instances.

If we implement it with these global variables, it would lead to one having to understanding the implementation details of this component to use them properly.

Given the folllowing snippet:
```
matcherA = php.NewTodomatcher("TODO", "todo")
matcherB = php.NewTodomatcher("FIXME", "fixme")
```

`matcherA` will now work with `FIXME` and `fixme` instead of the formats it was initialised with.

Reply:
I don't understand what you mean with this words. I have never seen or done this before. Can you give me links for read about this? Or describe little more please. This is my first language with strict types and I have never worked with factories before. So I don't understand how to transfer this optimization from matchers to the factory. I would be very grateful for your help.

Reply:
> Instead of making this optimization here, you can do it in the matcher factories. There, instead of initializing a matcher every time, initialise them as fields of the `matcherFactory` and reuse the instances.

I thought for all night about the problem described and I think I understood what you wrote about. Is it a correct solution?

```go
var lock sync.Mutex

type matcherFactory struct {
	newTodoMatcher     func(string) TodoMatcher
	newCommentsMatcher func(callback state.CommentCallback) CommentMatcher
	todoInstance       TodoMatcher
}

var standardMatcherFactory = &matcherFactory{
	func(customTodos string) TodoMatcher {
		if standardMatcherFactory.todoInstance == nil {
			lock.Lock()
			standardMatcherFactory.todoInstance = standard.NewTodoMatcher(customTodos)
			lock.Unlock()
		}
		return standardMatcherFactory.todoInstance
	},
	func(callback state.CommentCallback) CommentMatcher {
		return standard.NewCommentMatcher(callback, false)
	},
}
```

Reply:
yes, like this but no need for the lazy initialization.

Simply initialise it like so:
```go
var standardMatcherFactory = &matcherFactory{
        todoMatcher: standard.NewTodoMatcher()
	func(customTodos string) TodoMatcher {
		return standardMatcherFactory.todoMatcher
	},
	func(callback state.CommentCallback) CommentMatcher {
		...
	},
}
```

The `customTodos` need not be provided here. Provide them in the instance functions of the todo matcher. This avoids creating a state variable.

Reply:
Sorry in advance for stupid questions and inferences. I guess I don't understand what it should look like at all.

> The `customTodos` need not be provided here. Provide them in the instance functions of the todo matcher. This avoids creating a state variable.

Is this mean that I need add function like `setTodos` for matchers? When this function must be called?

I don't understand why this change will make the code clearer. There will be state variables with regular expressions in any matcher. This will make it possible to change sets of regular expressions at runtime. The configuration is set once at startup and is not changed at runtime, so I don't see the point in this method.

Reply:
hmm, okay. You're right that the matchers will have the regexps as state variables & if we provide the `customTodos` on each call to `IsMatch` and `IsValid`, then the regular expressions will need to be reinitialized.

In this case, this approach is ok with a suggestion for improving the synchronization like this:
```go
var (
	once            *sync.Once
	standardMatcher TodoMatcher
)

var (
	standardMatcherFactory = &matcherFactory{
		func(customTodos string) TodoMatcher {
			once.Do(func() {
				standardMatcher = standard.NewTodoMatcher(customTodos)
			})

			return standardMatcher
		},
		func(callback state.CommentCallback) CommentMatcher {
			return standard.NewCommentMatcher(callback, false)
		},
	}
        ...
```

Reply:
Optimized. Thank you for your advice.